# Tutorial 1.2 - Stochastic Gene Expression

Here we examine a CME model of stochastic Gene Expression.

In this model, we include the transcription and translation of gene and mRNA together with degradation of both mRNA and protein.

The model presented here can be found in the article: [Analytical distributions for stochastic gene expression](https://www.pnas.org/doi/full/10.1073/pnas.0803850105).


In [1]:
# Import Standard Python Libraries
import os
import numpy as np
import matplotlib.pyplot as plt

# Import pyLM Libraries
from pyLM import *
from pyLM.units import *
from pySTDLM import *
from pySTDLM.PostProcessing import *

# Enable plotting inline in the Jupyter notebook
%matplotlib inline

## Constants

Rates of first three first-order reactions come from the CME/ODE Whole Cell Model at initial conditions.

Degradation rate of protein is calculated based on 25 hours' half life in [Maier et al, 2011](https://www.embopress.org/doi/full/10.1038/msb.2011.38).

In [2]:
# Constants
v0  = 6.41e-4       # Transcription, s^-1
d0 = 2.59e-3     # degradation of mRNA, s^-1
v1 = 7.2e-2        # translation, s^-1
d1 = 7.70e-6      # degradation of protein, s^-1

## Define CME simulation

We begin by creating a [CMESimulation](https://luthey-schulten.chemistry.illinois.edu/software/LM2.4/_autosummary/pyLM.CME.html#module-pyLM.CME) "object" that we call ```sim```. This object will include the definition of the whole stochastic simulation.

In [3]:
# Create our CME simulation object
sim = CME.CMESimulation(name='Gene Expression')

Next we define the chemical species with simulation. First. we specify the names of the chemical species.  Then we register these species with the simulation.  The [```defineSpecies()```](https://luthey-schulten.chemistry.illinois.edu/software/LM2.4/_autosummary/pyLM.CME.html#module-pyLM.CME) function can be called multiple times and will add any new names to the list of species.

In [4]:
# Define our chemical species
species = ['gene', 'mRNA', 'ptn']
sim.defineSpecies(species)

Here we add reactions to the simulation. We use the [```addReaction()```](https://luthey-schulten.chemistry.illinois.edu/software/LM2.4/_autosummary/pyLM.CME.html#module-pyLM.CME) function that is a member of the ```CMESimulation``` object. We add a bimolecular association reaction and a unimolecular dissociation reaction. When more than one reactant is involved, the list of reactant names should be passed as a tuple as can be seen in the reactant of the association reaction. The rates in this command must be in units of molecules and seconds, for instance units of ```/molecule/sec``` for the association reaction.

In [5]:
# Add reactions to the simulation

sim.addReaction(reactant='gene', product=('gene','mRNA'), rate=v0)
sim.addReaction(reactant='mRNA', product='', rate=d0)
sim.addReaction(reactant='mRNA', product=('mRNA','ptn'), rate=v1)
sim.addReaction(reactant='ptn', product='', rate=d1)


Next, we add the initial particle counts to the simulation using the [```addParticles()```](https://luthey-schulten.chemistry.illinois.edu/software/LM2.4/_autosummary/pyLM.CME.html#module-pyLM.CME) function.

In [6]:
# Set our initial species counts

sim.addParticles(species='gene', count=1)
sim.addParticles(species='mRNA', count=1)
sim.addParticles(species='ptn', count=0)

Finally, we define the simulation execution parameters. We have the simulation run for 6300 seconds of real time to cover the entire cell cyle.

The traces are recorded per 1 second.

Then we name the simulation output file and save the simulation definition to it.

In [7]:
# Simulation time is 6300, entire cell life cycle.
writeInterval = 1
simtime = 6300

sim.setWriteInterval(writeInterval)
sim.setSimulationTime(simtime)

filename = "./T2.1-geneExpression.lm"

os.system("rm -rf %s"%(filename)) # Remove previous LM file 

sim.save(filename)

In [8]:
# Print out the information of the system
sim

## Run Simulation

In [9]:
# Run multipled replicates using the Gillespie solver
# Will cost less than 1 minute
reps = 200

sim.run(filename=filename, method="lm::cme::GillespieDSolver", replicates=reps)

  0%|                                                                                                                                                                              | 0/200 [00:00<?, ?it/s]

2024-04-29 12:10:19) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:19) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:19) Info: Data output thread running.
2024-04-29 12:10:19) Info: Seeding xorwow rng with top word 1 and bottom word 878123622
2024-04-29 12:10:19) Info: Data output thread finished.
2024-04-29 12:10:19) Info: Simulation file closed.


  0%|8                                                                                                                                                                     | 1/200 [00:00<00:28,  6.94it/s]

2024-04-29 12:10:19) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:19) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:19) Info: Data output thread running.
2024-04-29 12:10:19) Info: Seeding xorwow rng with top word 2 and bottom word 911345115
2024-04-29 12:10:19) Info: Data output thread finished.
2024-04-29 12:10:19) Info: Simulation file closed.
2024-04-29 12:10:19) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:19) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:19) Info: Data output thread running.
2024-04-29 12:10:19) Info: Seeding xorwow rng with top word 3 and bottom word 937386960
2024-04-29 12:10:19) Info: Data output thread finished.
2024-04-29 12:10:19) Info: Simulation file closed.
2024-04-29 12:10:19) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:19) Info: Assigning 1.00 processor(s) and 1.00

  2%|####1                                                                                                                                                                 | 5/200 [00:00<00:08, 23.13it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 6 and bottom word 14240195
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 7 and bottom word 39880141
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 C

  4%|#######4                                                                                                                                                              | 9/200 [00:00<00:06, 29.57it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 10 and bottom word 118033109
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 11 and bottom word 141449345
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

  7%|###########5                                                                                                                                                         | 14/200 [00:00<00:05, 34.38it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 15 and bottom word 239111795
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 16 and bottom word 263623830
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

  9%|##############8                                                                                                                                                      | 18/200 [00:00<00:05, 36.20it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 19 and bottom word 338954792
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 20 and bottom word 363681425
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

 12%|##################9                                                                                                                                                  | 23/200 [00:00<00:04, 38.50it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 24 and bottom word 456650931
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 25 and bottom word 480076241
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

 14%|#######################1                                                                                                                                             | 28/200 [00:00<00:04, 40.12it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 29 and bottom word 571460212
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 30 and bottom word 594910693
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

 16%|###########################2                                                                                                                                         | 33/200 [00:00<00:04, 41.01it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 34 and bottom word 688217264
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 35 and bottom word 711871678
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

 19%|###############################3                                                                                                                                     | 38/200 [00:01<00:03, 41.61it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 39 and bottom word 804710392
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 40 and bottom word 827611409
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

 22%|###################################4                                                                                                                                 | 43/200 [00:01<00:03, 42.27it/s]

2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 44 and bottom word 918780350
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:20) Info: Data output thread running.
2024-04-29 12:10:20) Info: Seeding xorwow rng with top word 45 and bottom word 942088615
2024-04-29 12:10:20) Info: Data output thread finished.
2024-04-29 12:10:20) Info: Simulation file closed.
2024-04-29 12:10:20) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:20) Info: Assigning 1.00 processor(s) and 1.

 24%|#######################################6                                                                                                                             | 48/200 [00:01<00:03, 42.30it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 49 and bottom word 37950986
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 50 and bottom word 62678763
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00

 26%|###########################################7                                                                                                                         | 53/200 [00:01<00:03, 42.09it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 54 and bottom word 157605098
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 55 and bottom word 182796447
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.

 29%|###############################################8                                                                                                                     | 58/200 [00:01<00:03, 41.57it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 59 and bottom word 281055015
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 60 and bottom word 308165004
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.

 32%|###################################################9                                                                                                                 | 63/200 [00:01<00:03, 40.31it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 64 and bottom word 413119130
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 65 and bottom word 436713389
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.

 34%|########################################################1                                                                                                            | 68/200 [00:01<00:03, 40.71it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 69 and bottom word 533522846
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 70 and bottom word 557064669
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.

 36%|############################################################2                                                                                                        | 73/200 [00:01<00:03, 41.17it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 74 and bottom word 651872182
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 75 and bottom word 676477790
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.

 39%|################################################################3                                                                                                    | 78/200 [00:02<00:02, 41.02it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 79 and bottom word 774583164
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 80 and bottom word 797804013
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.

 42%|####################################################################4                                                                                                | 83/200 [00:02<00:02, 41.16it/s]

2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 84 and bottom word 895123510
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:21) Info: Data output thread running.
2024-04-29 12:10:21) Info: Seeding xorwow rng with top word 85 and bottom word 918434477
2024-04-29 12:10:21) Info: Data output thread finished.
2024-04-29 12:10:21) Info: Simulation file closed.
2024-04-29 12:10:21) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:21) Info: Assigning 1.00 processor(s) and 1.

 44%|########################################################################6                                                                                            | 88/200 [00:02<00:02, 41.42it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 89 and bottom word 14233299
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 90 and bottom word 38278297
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00

 46%|############################################################################7                                                                                        | 93/200 [00:02<00:02, 41.46it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 94 and bottom word 134302727
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 95 and bottom word 156989746
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.

 49%|################################################################################8                                                                                    | 98/200 [00:02<00:02, 42.09it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 99 and bottom word 249077783
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 100 and bottom word 272896118
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1

 52%|####################################################################################4                                                                               | 103/200 [00:02<00:02, 40.84it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 104 and bottom word 379916980
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 105 and bottom word 406295430
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 

 54%|########################################################################################5                                                                           | 108/200 [00:02<00:02, 39.63it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 109 and bottom word 515326262
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 110 and bottom word 540606753
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 

 56%|###########################################################################################8                                                                        | 112/200 [00:02<00:02, 39.68it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 113 and bottom word 615082877
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 114 and bottom word 638596157
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 

 58%|###############################################################################################9                                                                    | 117/200 [00:02<00:02, 40.08it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 118 and bottom word 737629660
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 119 and bottom word 762000423
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 

 61%|####################################################################################################                                                                | 122/200 [00:03<00:01, 40.03it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 123 and bottom word 862559722
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 124 and bottom word 887251207
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 

 64%|########################################################################################################1                                                           | 127/200 [00:03<00:01, 40.09it/s]

2024-04-29 12:10:22) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:22) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:22) Info: Data output thread running.
2024-04-29 12:10:22) Info: Seeding xorwow rng with top word 128 and bottom word 986570143
2024-04-29 12:10:22) Info: Data output thread finished.
2024-04-29 12:10:22) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 129 and bottom word 11294935
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1

 66%|############################################################################################################2                                                       | 132/200 [00:03<00:01, 40.56it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 133 and bottom word 106868429
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 134 and bottom word 130554083
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 

 68%|################################################################################################################3                                                   | 137/200 [00:03<00:01, 40.97it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 138 and bottom word 226035568
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 139 and bottom word 249316886
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 

 71%|####################################################################################################################4                                               | 142/200 [00:03<00:01, 41.15it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 143 and bottom word 346197106
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 144 and bottom word 371573472
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 

 74%|########################################################################################################################5                                           | 147/200 [00:03<00:01, 41.06it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 148 and bottom word 468635398
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 149 and bottom word 494351510
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 

 76%|############################################################################################################################6                                       | 152/200 [00:03<00:01, 40.90it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 153 and bottom word 591803748
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 154 and bottom word 617967951
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 

 78%|################################################################################################################################7                                   | 157/200 [00:03<00:01, 40.85it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 158 and bottom word 715755694
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 159 and bottom word 742615093
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 

 81%|####################################################################################################################################8                               | 162/200 [00:04<00:00, 39.44it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 163 and bottom word 851898058
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 164 and bottom word 876741743
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 

 84%|########################################################################################################################################9                           | 167/200 [00:04<00:00, 39.96it/s]

2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:23) Info: Seeding xorwow rng with top word 168 and bottom word 972891716
2024-04-29 12:10:23) Info: Data output thread finished.
2024-04-29 12:10:23) Info: Simulation file closed.
2024-04-29 12:10:23) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:23) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:23) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 169 and bottom word 998440390
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 

 86%|#############################################################################################################################################                       | 172/200 [00:04<00:00, 39.74it/s]

2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 173 and bottom word 100487557
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 174 and bottom word 124830433
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 

 88%|#################################################################################################################################################1                  | 177/200 [00:04<00:00, 40.13it/s]

2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 178 and bottom word 222002573
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 179 and bottom word 245063985
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 

 91%|#####################################################################################################################################################2              | 182/200 [00:04<00:00, 40.57it/s]

2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 183 and bottom word 342299361
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 184 and bottom word 366085082
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 

 94%|#########################################################################################################################################################3          | 187/200 [00:04<00:00, 41.13it/s]

2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 188 and bottom word 459853214
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 189 and bottom word 483721857
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 

 96%|#############################################################################################################################################################4      | 192/200 [00:04<00:00, 41.43it/s]

2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 193 and bottom word 578448234
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 194 and bottom word 603531922
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 

 98%|#################################################################################################################################################################5  | 197/200 [00:04<00:00, 41.28it/s]

2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 198 and bottom word 702232153
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 1.00 CUDA device(s) per replicate.
2024-04-29 12:10:24) Info: Data output thread running.
2024-04-29 12:10:24) Info: Seeding xorwow rng with top word 199 and bottom word 725417191
2024-04-29 12:10:24) Info: Data output thread finished.
2024-04-29 12:10:24) Info: Simulation file closed.
2024-04-29 12:10:24) Info: Using 56 processor(s) and 1 CUDA device(s) per process.
2024-04-29 12:10:24) Info: Assigning 1.00 processor(s) and 

100%|####################################################################################################################################################################| 200/200 [00:05<00:00, 39.95it/s]


## Post-Processing

In [10]:
fileHandle = PostProcessing.openLMFile(filename) # Create h5py file handle

plotfolder = './plots_gene_expression/'

if not os.path.exists(plotfolder):
    os.mkdir(plotfolder)


#### Using [```plotAvgVarFromFile()```](https://luthey-schulten.chemistry.illinois.edu/software/LM2.4/_autosummary/pySTDLM.PostProcessing.html#module-pySTDLM.PostProcessing) built-in function in PostProcessing to plot the average and variance of species. 

In [11]:
for i_specie, specie in enumerate(species):

    picturepath = plotfolder + 'Trace.{0}.png'.format(specie)

    PostProcessing.plotAvgVarFromFile(filename = filename, species = [specie], outfile = picturepath)

<Figure size 640x480 with 0 Axes>

#### Using user-defined function to serialize traces in LM file to a 3D Numpy Array with dimesions *(reps, species, time)*.

In [12]:
import h5py

timestep = PostProcessing.getTimesteps(fileHandle) # use PostProcessing to get the timesteps of the simulation

traces = np.zeros((reps,len(sim.particleMap),len(timestep)))

def get_sim_data(filename):

    f=h5py.File(filename, "r")

    for r in range(reps):
        
        traces[r]=f['Simulations'][str(r+1).zfill(7)]['SpeciesCounts'][()].transpose()
        
    f.close()

    return traces

traces = get_sim_data(filename)

# print(traces)

#### Plot the distribution of Protein at the end of the simulation

User defined function to draw histogram

In [13]:
def plot_histogram(data, figure_path, bins, xlabel, title):
    
    fig_size = [87,87/1.618]

    fig = plt.figure(figsize=(fig_size[0]/25.4,fig_size[1]/25.4))

    plt.hist(ptns_end, bins=20, alpha=0.7, color='limegreen')

    ax = plt.gca()

    xlabel = xlabel.replace('_','\_')
    ax.set_xlabel(r'{0}'.format(xlabel),
                  fontsize=7,
                  labelpad=1.5)

    ax.set_ylabel(r'{0}'.format('Frequency'),
                fontsize=7,
                labelpad=1.5)

    title = title.replace('_','\_')
    ax.set_title(r'{0}'.format(title),
                 fontsize=8,
                 pad=4)

    tick_length = 4.0
    tick_width = 1.5
    ax.tick_params(labelsize=5,
                    length=tick_length,
                    width=tick_width,
                    direction='in',
                    left=True,
                    right=True,
                    bottom=True,
                    top=True,
                    which='major')

    ax.tick_params(labelsize=5,
                    length=tick_length/1.5,
                    width=tick_width/1.5,
                    direction='in',
                    left=True,
                    right=False,
                    bottom=True,
                    top=False,
                    which='minor')

    ax.spines['left'].set_linewidth(1.5)
    ax.spines['bottom'].set_linewidth(1.5)
    ax.spines['right'].set_linewidth(1.5)
    ax.spines['top'].set_linewidth(1.5)

    mean = np.mean(ptns_end)
    plt.axvline(mean, color='red', linestyle='dashed', linewidth=1.5, label = 'Mean: {0:.3f}'.format(mean))

    median = np.median(ptns_end)
    plt.axvline(median, color='black', linestyle='dashed', linewidth=1.5, label = 'Median: {0:.3f}'.format(median))      

    min = np.min(ptns_end)
    plt.axvline(min, color='blue', linestyle='dashed', linewidth=1.5, label = 'Min: {0:.3f}'.format(min))  

    ax.legend(fontsize = 8)

    fig.savefig(fig_path, dpi = 300)

    plt.close()
    
    return None

In [14]:
ptns_end = traces[:,2,-1] # Slice Numpy array to get the counts of ptn at the end of the whole cell cycle
# print(ptns_end)

fig_path = plotfolder + 'Distribution_Ptns_End.png'

xlabel = 'Counts of Protein [#]'

title = 'Distribution of Ptn Counts at the end of the simulation'
        
plot_histogram(data = ptns_end, figure_path = fig_path, bins = 20, xlabel = xlabel, title = title)

# plt.hist(ptns_end)

In [15]:
ptns_end = traces[:,1,-1] # Slice Numpy array to get the counts of ptn at the end of the whole cell cycle
# print(ptns_end)

fig_path = plotfolder + 'Distribution_mRNA_End.png'

xlabel = 'Counts of mRNA [#]'

title = 'Distribution of mRNA Counts at the end of the simulation'
        
plot_histogram(data = ptns_end, figure_path = fig_path, bins = 5, xlabel = xlabel, title = title)

In [16]:
break

SyntaxError: 'break' outside loop (668683560.py, line 4)

In [ ]:
from scipy.special import gamma

a = v0/d1
b = v1/d0

print('a',a,'b',b)
ns = np.linspace(0, 250, 251)

prefactor = gamma(ns+a)/gamma(ns+1)/gamma(a)

print(prefactor)

In [ ]:
import math as m

m.gamma(3000)

## Questions 2.1

1. Do mRNA and protein reach steady-state during the entire cell cycle? How can you tell this from the plots?

2. In the whole cell model of JCVI-SYN3A, the initial count of protein (P_0001/DnaA) is 148. Does the protein count double during the entire cell cyle? And why this is important?

3. (Challenge) Try to increase the simulation time to see the steady state of protein.

